File: ChessMetric.java
public class ChessMetric {
  public long howMany(int size, int[] start, int[] end, int numMoves) {
    int[][] dist = new int[][] { { -1, 0 }, { 1, 0 }, { 0, -1 }, { 0, 1 }, { 1, 1 }, { 1, -1 }, { -1, 1 }, { -1, -1 },
        { -2, -1 }, { -2, 1 }, { 2, -1 }, { 2, 1 }, { -1, -2 }, { 1, -2 }, { -1, 2 }, { 1, 2 }, };
    long[][][] ans = new long[numMoves + 1][size][size];

    ans[0][start[0]][start[1]] = 1;

    for (int i = 0; i < numMoves; i++) {
      for (int y = 0; y < size; y++) {
        for (int x = 0; x < size; x++) {
          if (ans[i][y][x] == 0)
            continue;

          for (int di = 0; di < 16; di++) {
            int ny = y + dist[di][0];
            int nx = x + dist[di][1];

            if (ny < 0 || ny >= size || nx < 0 || nx >= size)
              continue;

            ans[i + 1][ny][nx] += ans[i][y][x];
          }
        }
      }
    }

    return ans[numMoves][end[0]][end[1]];
  }
}

File: ChessMetric.cpp
#include <algorithm>
#include <bitset>
#include <cassert>
#include <cmath>
#include <ctime>
#include <functional>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <vector>

#define REP(i,n)   for (int i = 0; i < (n); ++i)
#define SORT(a)    sort((a).begin(), (a).end());
#define UNIQ(a)    SORT(a);(a).erase(unique((a).begin(), (a).end()), (a).end());
#define FIND(a,x)  find((a).begin(), (a).end(), (x)) != (a).end()
#define REVERSE(a) reverse((a).begin(), (a).end());

using namespace std;
using ll = long long;
using P = pair<int,int>;

const int MOD = 1000000007; // 10^9 + 7

const int MAX_SIZE = 105;
const int MAX_MOVES = 50;
const int dx[16] = {0, 1, 1, 1, 0, -1, -1, -1, 1, 2, 2, 1, -1, -2, -2, -1};
const int dy[16] = {-1, -1, 0, 1, 1, 1, 0, -1, -2, -1, 1, 2, 2, 1, -1, -2};
int SIZE;
ll dp[MAX_SIZE][MAX_SIZE][MAX_MOVES];

class ChessMetric {
  public:
  ll howMany(int size, vector <int> start, vector <int> end, int numMoves) {
    SIZE = size;
    REP(i, MAX_SIZE)REP(j, MAX_SIZE)REP(k, MAX_MOVES) dp[i][j][k] = -1;

    dp[start[0]][start[1]][0] = 1;

    return solve(end[0], end[1], numMoves);
  }

  private:
  ll solve(int x, int y, int move) {
    // cout << "x = " << x << ", y = " << y << ", move = " << move << endl;
    if (dp[x][y][move] != -1) return dp[x][y][move];
    if (move == 0) return 0;

    ll ret = 0;
    REP(i, 16) {
      int nx = x - dx[i];
      int ny = y - dy[i];
      if (nx >= 0 && nx < SIZE && ny >= 0 && ny < SIZE) {
        ret += solve(nx, ny, move - 1);
      }
    }

    return dp[x][y][move] = ret;

    return 0;
  }
};

int main() {
  cin.tie(0);
  ios::sync_with_stdio(false);

  ChessMetric *cm = new ChessMetric;
  cout << cm->howMany(3, {0, 0}, {1, 0}, 1) << endl;
}